File: python/verify.py
import sys
import jwt
import json

# Command-line arguments
if len(sys.argv) < 3:
    print("Usage: python verify.py <jwtFile> <publicKeyFile>")
    sys.exit(1)

jwt_file = sys.argv[1]
public_key_file = sys.argv[2]

# Read the JWT
try:
    with open(jwt_file, 'r') as f:
        token = f.read().strip()
except FileNotFoundError:
    print(f"Error: JWT file '{jwt_file}' not found.")
    sys.exit(1)

# Read the public key
try:
    with open(public_key_file, 'r') as f:
        public_key = f.read()
except FileNotFoundError:
    print(f"Error: Public key file '{public_key_file}' not found.")
    sys.exit(1)

# Decode without verification
try:
    decoded_unverified = jwt.decode(token, options={"verify_signature": False}, algorithms=["ES256"])
    print("\n", json.dumps(decoded_unverified, indent=4))
except Exception as e:
    print(f"Error decoding JWT without verification: {e}")

# Decode and verify the JWT
try:
    decoded = jwt.decode(token, public_key, algorithms=["ES256"], options={"verify_exp": False})
    print("Verification Succeeded")
except jwt.ExpiredSignatureError:
    print("Error: JWT signature has expired.")
except jwt.InvalidSignatureError:
    print("Error: JWT signature is invalid.")
except jwt.DecodeError:
    print("Error: Failed to decode JWT.")
except Exception as e:
    print(f"Error: {e}")

File: dotnet/verify.cs
using System;
using System.IO;
using System.IdentityModel.Tokens.Jwt;
using Microsoft.IdentityModel.Tokens;
using System.Text;
using System.Text.Json;

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: dotnet run <jwtFile> <publicKeyFile>");
            return;
        }

        string jwtFile = args[0];
        string publicKeyFile = args[1];

        // Read the JWT
        string token;
        try
        {
            token = File.ReadAllText(jwtFile).Trim();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error reading JWT file: {ex.Message}");
            return;
        }

        // Read the public key
        string publicKeyPem;
        try
        {
            publicKeyPem = File.ReadAllText(publicKeyFile);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error reading public key file: {ex.Message}");
            return;
        }

        // Convert PEM to RSA public key
        var publicKey = PemToEcdsaSecurityKey(publicKeyPem);

        // Verify the JWT
        var validationParameters = new TokenValidationParameters
        {
            ValidateIssuer = false,
            ValidateAudience = false,
            ValidateLifetime = false,
            IssuerSigningKey = publicKey
        };

        var handler = new JwtSecurityTokenHandler();
        // Decode JWT without verification
        try
        {
            var jwt = handler.ReadJwtToken(token);
            Console.WriteLine($"{jwt.Header.SerializeToJson()}");
            Console.WriteLine($"{PrettyPrintJson(jwt.Payload.SerializeToJson())}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error decoding JWT: {ex.Message}");
        }
        try
        {
            var claims = handler.ValidateToken(token, validationParameters, out _);
            Console.WriteLine("Verification Succeeded");
            // foreach (var claim in claims.Claims)
            // {
            //     Console.WriteLine($"{claim.Type}: {claim.Value}");
            // }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Verification Failed: {ex.Message}");
        }        
    }
    // Helper method to pretty print JSON
    private static string PrettyPrintJson(string json)
    {
        var options = new JsonSerializerOptions
        {
            WriteIndented = true, // Enable pretty printing
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping // Avoid escaping +
        };
        var jsonElement = JsonSerializer.Deserialize<JsonElement>(json);
        return JsonSerializer.Serialize(jsonElement, options);
    }
    // Helper method to convert PEM to ECDSA Security Key
    private static ECDsaSecurityKey PemToEcdsaSecurityKey(string pem)
    {
        var ecdsa = System.Security.Cryptography.ECDsa.Create();
        ecdsa.ImportFromPem(pem);
        return new ECDsaSecurityKey(ecdsa);
    }
}
